test/insert.js

```diff
@@ -0,0 +1,57 @@
+'use strict';
+
+var S = require('..');
+
+var eq = require('./internal/eq');
+
+
+test('insert', function() {
+
+  eq(typeof S.insert, 'function');
+  eq(S.insert.length, 3);
+  eq(S.insert.toString(), 'insert :: String -> a -> StrMap a -> StrMap a');
+
+  function testBasic(key, val, strMap, strMapResult) {
+    eq(S.insert(key, val, strMap), strMapResult);
+  }
+
+  function unique(sortedArray) {
+    var uniqueArray = [];
+    sortedArray.forEach(function(val, i) {
+      if (i === 0 || val !== sortedArray[i - 1]) {
+        uniqueArray.push(val);
+      }
+    });
+    return uniqueArray;
+  }
```

Comment:
Here's a less imperative implementation:

```javascript
function unique(xs) {
  if (xs.length === 0) return [];

  var sorted = S.sort(xs);
  var x = sorted[0];
  return sorted.slice(1).reduce(function(acc, x) {
    return S.equals(x, acc.prev) ? acc : {xs: S.append(x, acc.xs), prev: x};
  }, {xs: [x], prev: x}).xs;
}
```

I think we should avoid requiring a sorted array as input, even if this means we end up sorting twice in some cases.

Reply:
What about:
```javascript
function unique(xs) {
  return S.sort(xs).reduce(function(acc, x, i, xs) {
    return i > 0 && S.equals(x, xs[i-1]) ? acc : S.append(x, acc)
  }, []);
}
```
or
```javascript
function appendIfUnique(acc, x, i, xs) {
  return i > 0 && S.equals(x, xs[i-1])
    ? acc
    : S.append(x, acc);
}

function unique(xs) {
  return S.sort(xs).reduce(appendIfUnique, []);
}
```

Reply:
The use of indexes and the `i > 0` guard in particular are the things which bother me. If we're to deal with indexes I'd prefer to use a `for` loop:

```javascript
function unique(xs) {
  if (xs.length === 0) return [];

  var sorted = S.sort(xs);
  var prev = sorted[0];
  var result = [prev];
  for (var idx = 1; idx < sorted.length; idx += 1) {
    if (!S.equals(sorted[idx], prev)) result.push(prev = sorted[idx]);
  }
  return result;
}
```

Reply:
I can replace with your `reduce` implementation. Just for my own edification, curious why you don't like indexes and `i > 0`? Is it the potential for getting the wrong array element?

Reply:
I'd prefer to handle the special case (`[]`) in one place rather than allow it to complicate the common code path. I don't mind using array indexes when implementing Sanctuary functions, but in contexts in which performance is less of a concern (such as a test suite) I'd prefer to avoid "cheating" (keep in mind that `S.map` and friends do not expose indexes). In this case we're using an index to access the previous element, but it seems more natural to me to keep track of this.

This is all subjective, and I don't mean to assert that my preferences are better than alternatives in any meaningful way.

Reply:
Updated pull request.